R/elasticsearch_parsers.R

```diff
@@ -402,46 +403,58 @@ unpack_nested_data <- function(chomped_df, col_to_unpack) {
         stop(msg)
     }
     
-    # Avoid side effects
     outDT <- data.table::copy(chomped_df)
-    
-    # Get the column to unpack
     listDT <- outDT[[col_to_unpack]]
     
-    # Make each row a data.table
-    listDT <- lapply(listDT, data.table::as.data.table)
-    
-    # Remove the empty ones... important, due to data.table 1.10.4 bug
-    oldIDs <- which(sapply(listDT, nrow) != 0)
-    listDT <- listDT[oldIDs]
-    
-    # Bind them together with an ID to match to the other data
-    newDT <- data.table::rbindlist(listDT, fill = TRUE, idcol = TRUE)
-    
-    # If we tried to unpack an empty column, fail
-    if (nrow(newDT) == 0) {
+    # Check for empty column
+    if (all(lengths(listDT) == 0)) {
         msg <- "The column given to unpack_nested_data had no data in it."
         futile.logger::flog.fatal(msg)
         stop(msg)
     }
     
-    # Fix the ID because we may have removed some empty elements due to that bug
-    newDT[, .id := oldIDs[.id]]
+    listDT[lengths(listDT) == 0] <- NA
     
-    # Merge
-    outDT[, .id := .I]
-    outDT <- newDT[outDT, on = ".id"]
+    is_df <- purrr::map_lgl(listDT, is.data.frame)
+    is_na <- is.na(listDT)
+    is_atomic <- purrr::map_lgl(listDT, purrr::is_atomic)
     
-    # Remove the id column and the original column
-    outDT <- outDT[, !c(".id", col_to_unpack), with = FALSE]
+    # Bind packed column into one data.table
+    if (all(is_df | is_na)) {
+        # If the packed column contained data.frames, we need to create a row of 
+        # NAs data.table for each empty row in the packed column
+        first_nonmissing <- min(which(is_df))
+        DTNames <- names(listDT[[first_nonmissing]])
+        
+        newDT <- purrr::map(
+            listDT, 
+            function(x) {
+                if(!is.data.frame(x)) {
+                    x <- as.list(setNames(rep(NA, length(DTNames)), DTNames))
+                    x <- data.table::as.data.table(x)
+                }
+                x
+            }
+        )
+        newDT <- data.table::rbindlist(newDT, fill = TRUE)
+    } else if (all(is_atomic)) {
+        newDT <- data.table::as.data.table(unlist(listDT))
+    } else {
+        msg <- "For unpack_nested_data, col_to_unpack must be all atomic vectors or all data frames"
```

Comment:
I'd like if we gave a more informative error message here. If I got this error as a user I wouldn't know what to do next.

Please do two things:
1) write a unit test using `expect_error()` that reaches this line
2) propose an error message format that might be able to give a user a bit more guidance on where to go next